File: core/src/main/java/uk/co/amlcurran/social/CalendarSource.java
package uk.co.amlcurran.social;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import uk.co.amlcurran.social.core.SparseArray;

class CalendarSource {

    private final SparseArray<CalendarSlot> calendarItems;
    private final int daysSize;
    private final TimeCalculator timeCalculator;

    public CalendarSource(SparseArray<CalendarSlot> calendarItems, int daysSize, TimeCalculator timeCalculator) {
        this.calendarItems = calendarItems;
        this.daysSize = daysSize;
        this.timeCalculator = timeCalculator;
    }

    public int count() {
        return daysSize;
    }

    @Nullable
    public CalendarItem itemAt(int position) {
        CalendarSlot calendarSlot = calendarItems.get(position);
        if (calendarSlot == null || calendarSlot.isEmpty()) {
            Timestamp startTime = startOfTodayBlock(position);
            Timestamp endTime = endOfTodayBlock(position);
            return new EmptyCalendarItem(startTime, endTime);
        }
        return calendarSlot.firstItem();
    }

    @Nonnull
    private Timestamp startOfTodayBlock(int position) {
        return timeCalculator.startOfToday().plusDays(position).plusHours(17);
    }

    @Nonnull
    private Timestamp endOfTodayBlock(int position) {
        return timeCalculator.startOfToday().plusDays(position).plusHours(23);
    }

    @Nonnull
    public CalendarSlot slotAt(int position) {
        return calendarItems.get(position, new CalendarSlot());
    }

    public boolean isEmptySlot(int position) {
        return slotAt(position).isEmpty();
    }
}

File: android/shared/src/commonMain/kotlin/uk/co/amlcurran/social/CalendarSource.kt
package uk.co.amlcurran.social

import kotlinx.datetime.Instant

data class CalendarSource(private val calendarItems: Map<Int, CalendarSlot>, private val daysSize: Int, private val timeCalculator: TimeCalculator, private val userSettings: UserSettings) {

    private fun startOfTodayBlock(position: Int): Instant {
        return timeCalculator.startOfToday()
            .plusDays(position)
            .plusHoursOf(userSettings.borderTimeStart())
    }

    private fun endOfTodayBlock(position: Int): Instant {
        return timeCalculator.startOfToday()
            .plusDays(position)
            .plusHoursOf(userSettings.borderTimeEnd())
    }

    fun slotAt(position: Int): CalendarSlot {
        return calendarItems[position] ?: CalendarSlot(
            mutableListOf(),
            startOfTodayBlock(position),
            endOfTodayBlock(position)
        )
    }

    fun allSlots(): List<CalendarSlot> {
        return (0 until daysSize).map {
            slotAt(it)
        }
    }

    fun isEmptySlot(position: Int): Boolean {
        return slotAt(position).isEmpty
    }
}
